### PR TITLE
Add segyio-catr.c

### PR DESCRIPTION
**Pull request by ErlendHaa**
segyio-catr.c is a console based application that print a single, or a range, of trace header for a given SEG-Y file.
Added on request from segyio-users.

### CODE REVIEW

File: applications/CMakeLists.txt [hunk 1]
```diff
@@ -22,7 +22,16 @@ target_compile_definitions(segyio-cath PRIVATE
...
+    -Dsegyio_MAJOR=${segyio_MAJOR}
+    -Dsegyio_MINOR=${segyio_MINOR}
+)
+
 install(TARGETS segyinfo DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
 install(TARGETS segyinspect DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
 
 install(TARGETS segyio-cath DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+add_python_test(applications.apps   test/apps.py)
```
Comment (kjellkongsvik) [inline applications/CMakeLists.txt]: Add newline at end of file

File: applications/segyio-catr.c [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+    "smm"     ,
+    "sme"     ,
+    "smunit"  ,
+    "uint1"   ,
+    "uint2"
+};
+
+static int help() {
+    puts( "Usage: segyio-r [OPTION]... SRC\n"
+          "Print spesific traceheaders from SRC\n"
```
Comment (kjellkongsvik) [inline applications/segyio-catr.c]: typo: specific

File: applications/segyio-catr.c [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+#define TRHSIZE SEGY_TRACE_HEADER_SIZE
+#define BINSIZE SEGY_BINARY_HEADER_SIZE
+
+int main( int argc, char** argv) {
+
+    static int version = 0;
+    static int verbosity = 0;
+
+    static struct option long_options[] = {
+	{ "traceheader",		        required_argument, 	0,	    	't' },
```
Comment (kjellkongsvik) [inline applications/segyio-catr.c]: remove all unneeded white-space.

Comment (kjellkongsvik) [inline applications/segyio-catr.c]: at least remove all tabs

File: applications/segyio-catr.c [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+                }
+                printf("\n");
+                //printf("Trace: %d\n", i);
+            }
+        }
+        segy_close( src );
+    }
+    return 0;
+}
+
```
Comment (kjellkongsvik) [inline applications/segyio-catr.c]: Only one empty line at end

File: applications/test/apps.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        self.res_list = self.res.split(b"\n")
+        for i in range(len(self.ref_list)):
+            self.assertEquals(self.ref_list[i], self.res_list[i+92])
+
+    def test_catr_closed_interval_singel(self):
+        self.cmd = [self.cmd_base, self.work_dir, "-i 1", "-I 2", "-t 5"]
+        self.res = subprocess.Popen(self.cmd, stdout=subprocess.PIPE).communicate()[0]
+        self.res_list = self.res.split(b"\n")
+        for i in range(len(self.ref_list)):
+            self.assertEquals(self.ref_list[i], self.res_list[i+184])
```
Comment (kjellkongsvik) [inline applications/test/apps.py]: add empty line at end of file

File: applications/segyio-catr.c [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+   fputc( ' ', stderr );
+   return errmsg( errcode, msg );
+}
+
+static int printversion() {
+    printf( "Segyio-r (segyio version %d.%d)\n", segyio_MAJOR, segyio_MINOR);
+    return 0;
+}
+
+static void print_header( const char* header ) {
```
Comment (jokva) [inline applications/segyio-catr.c]: Unused.

File: applications/segyio-catr.c [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+   if( !prelude ) return errmsg( errcode, msg);
+
+   fputs( prelude, stderr );
+   fputc( ':', stderr );
+   fputc( ' ', stderr );
+   return errmsg( errcode, msg );
+}
+
+static int printversion() {
+    printf( "Segyio-r (segyio version %d.%d)\n", segyio_MAJOR, segyio_MINOR);
```
Comment (jokva) [inline applications/segyio-catr.c]: Reflect binary name.

File: applications/segyio-catr.c [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+          "Print specific traceheaders from SRC\n"
+	      "\n"
+          "-t, --traceheader=NUMBER                         Single trace header\n"
+          "-i, --traceheader-int-beg=NUMBER                 First trace header in interval\n"
+          "-I, --traceheader-int-end=NUMBER                 Last trace header in interval (inclusive)\n"
+          "-v, --verbose             		         Increase verbosity\n"
+          "    --version             		         Output version information and exit\n"
+          "    --help                		         Display this help and exit\n"
+          "\n"
+          "The < -t >tag can be set multiple times and \n"
```
Comment (jokva) [inline applications/segyio-catr.c]: No markup on flag description.

File: applications/segyio-catr.c [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+static int parseint( const char* str, int* x ) {
+    char* endptr;
+     *x = strtol( str, &endptr, 10);
+
+    if( *endptr != '\0' ) return 1;
+    if( *x < 0 ) return 2;
+    return 0;
+}
+
+static int samp_size( char const* bheader, int* samsize ) {
```
Comment (jokva) [inline applications/segyio-catr.c]: Can be removed in favour of `segy_trace_bsize`

File: applications/segyio-catr.c [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+
+    return 0;
+}
+
+#define TRHSIZE SEGY_TRACE_HEADER_SIZE
+#define BINSIZE SEGY_BINARY_HEADER_SIZE
+
+int main( int argc, char** argv) {
+
+    if( argc < 2 ) {
```
Comment (jokva) [inline applications/segyio-catr.c]: Check  this after parsing arguments.

File: applications/segyio-catr.c [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+	{ 0, 0, 0, 0,}
+    };
+
+    int trhno = -1;
+    int trhbeg = INT_MAX;
+    int trhend = -1;
+
+    static const char* parsenum_errmsg[] = { "", "num must be an integer",
+						 "num must be non-negative" };
+    while( true ) {
```
Comment (jokva) [inline applications/segyio-catr.c]: Move argument parsing to own function

File: applications/segyio-catr.c [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+    char trheader[ TRHSIZE ];
+    char binheader[ BINSIZE ];
+
+    for( int i = optind; i < argc; ++i ) {
+        segy_file* src = segy_open( argv[ i ], "r" );
+
+        if( !src )
+            return errmsg2( errno, "Unable to open src", strerror( errno ) );
+
+        if( argc == 2 ) {
```
Comment (jokva) [inline applications/segyio-catr.c]: Broken check.

File: applications/segyio-catr.c [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+            return help();
+        }
+
+        int err = segy_binheader(src, binheader);
+        if(err != 0)
+            return errmsg(errno, "Unable to read binheader");
+
+        int samnr;
```
Comment (jokva) [inline applications/segyio-catr.c]: Use `segy_trace_samples`

File: applications/segyio-catr.c [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+        err = segy_get_bfield(binheader, SEGY_BIN_SAMPLES, &samnr);
+        if(err != 0)
+            return errmsg(errno, "Unable to read binheader field: Samples");
+
+        int samsize;
```
Comment (jokva) [inline applications/segyio-catr.c]: `segy_trace_bsize`

File: applications/segyio-catr.c [hunk 15]
```diff
@@ -0,0 +1,426 @@
...
+
+        long trace0 = segy_trace0( binheader );
+
+        int numtrh;
+        err = segy_get_bfield( binheader, SEGY_BIN_TRACES, &numtrh );
+        if( err != 0 )
+            return errmsg(errno, "Unable to determine number of traces in file");
+
+        err = argument_range( trhno, trhbeg, trhend, numtrh );
+            if( err != 0 )
```
Comment (jokva) [inline applications/segyio-catr.c]: Offer option to not exit on this error condition

File: applications/segyio-catr.c [hunk 16]
```diff
@@ -0,0 +1,426 @@
...
+                return errmsg(errno, "Unable to read trace header");
+
+            int print_trh = check_args(trhno, trhbeg, trhend, i, numtrh);
+            if( print_trh == 0 ){
+                for( int i = 0; i < 91; i++ ) {
+                    int f;
+                    err = segy_get_field(trheader, fields[i], &f);
+                    if(err != 0)
+                        return errmsg(errno, "Unable to read trace field");
+                    printf("%s\t%i\n", su[i], f);
```
Comment (jokva) [inline applications/segyio-catr.c]: `%d`, not `%i`

**Comment by jokva** [comment]
Features added through other PR. Closing.

**Pull request closed without merging**